common/beerocks/bwl/shared/netlink_socket.cpp

```diff
@@ -0,0 +1,147 @@
+            }
+        });
+    LOG_IF(!nl_callback, FATAL) << "Failed creating netlink callback!";
+
+    // Termination flag for the loop that receives the response messages. Possible values are:
+    // error == 1: initial value, request message has not be sent yet
+    // error == 0: response has been successfully received
+    // error < 0: some error occurred while receiving response
+    // Final value is used to compute the result code of this method.
+    int error = 1;
+
+    // Create standard callbacks
+    static auto nl_err_cb = [](struct sockaddr_nl *nla, struct nlmsgerr *err, void *arg) -> int {
+        int *error = (int *)arg;
+        *error     = err->error;
+        LOG(ERROR) << "Failed to process netlink message! " << strerror(*error) << " (" << *error
+                   << ")";
+        return NL_STOP;
+    };
+
+    static auto nl_finish_cb = [](struct nl_msg *msg, void *arg) -> int {
+        int *error = (int *)arg;
+        *error     = 0;
+        return NL_SKIP;
+    };
+
+    static auto nl_ack_cb = [](struct nl_msg *msg, void *arg) -> int {
+        int *error = (int *)arg;
+        *error     = 0;
+        return NL_STOP;
+    };
+
+    // Passing a lambda with capture is not supported for standard C function
+    // pointers. As a workaround, we create a static (but thread local) wrapper
+    // function that calls the capturing lambda function.
+    static thread_local std::function<int(struct nl_msg * msg, void *arg)> nl_handler_cb_wrapper;
+    nl_handler_cb_wrapper = [&](struct nl_msg *msg, void *arg) -> int {
+        if (!msg_handle(msg)) {
+            LOG(ERROR) << "User's netlink handler function failed!";
+        }
+        return NL_SKIP;
+    };
+    auto nl_handler_cb = [](struct nl_msg *msg, void *arg) -> int {
+        return nl_handler_cb_wrapper(msg, arg);
+    };
+
+    // Call the user's message create function
+    if (!msg_create(nl_message.get())) {
+        LOG(ERROR) << "User's netlink create function failed!";
+        return false;
+    }
+
+    // Set the callbacks
+    nl_cb_err(nl_callback.get(), NL_CB_CUSTOM, nl_err_cb, &error); // error
+    nl_cb_set(nl_callback.get(), NL_CB_FINISH, NL_CB_CUSTOM, nl_finish_cb,
+              &error);                                                        // finish
+    nl_cb_set(nl_callback.get(), NL_CB_ACK, NL_CB_CUSTOM, nl_ack_cb, &error); // ack
+    nl_cb_set(nl_callback.get(), NL_CB_VALID, NL_CB_CUSTOM, nl_handler_cb,
+              nullptr); // response handler
+
+    // Send the netlink message
+    int rc = nl_send_auto_complete(m_nl_socket, nl_message.get());
+    if (rc < 0) {
+        LOG(ERROR) << "Failed to send netlink message! Error: " << rc;
+        return false;
+    }
+
+    // Receive the response messages
+    // Note that call to nl_recvmsgs() is blocking and loop terminates when one of the callback
+    // functions sets error to 0 (ok) or to a value lower than 0 (error)
+    while (error > 0) {
```

Comment:
I guess `error` should be of atomic_t type.

Reply:
Nope, it is set and read from the same context